nixos/modules/services/x11/xidlehook.nix

```diff
@@ -0,0 +1,225 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let cfg = config.services.xserver.xidlehook;
+in {
+  options = {
+    services.xserver.xidlehook = {
+      enable = mkEnableOption "xidlehook";
+      timers = mkOption {
+        description = ''
+          List of timers.
+
+          Most other xidlehook options adjust the default value of this option;
+          if this option is set directly, most other xidlehook options have no effect.
+          This is done to maintain drop-in compatibility with the xautolock module
+          for folks that already have xautolock set up the way they like it.
+          If you don't have an existing xautolock config, you may prefer the
+          simplicity of xidlehook's one-list-of-timers interface.
+       '';
+        example = [
+          # This is upstream's example: Dim the screen after 60 seconds, undim if user becomes active
+          {
+            duration = 60;
+            command = ''
+              ${pkgs.xorg.xrandr}/bin/xrandr --output "$(${pkgs.xorg.xrandr}/bin/xrandr | ${pkgs.gawk}/bin/awk '/primary/ { print $1 }')" --brightness .1
+            '';
+            canceller = ''
+              ${pkgs.xorg.xrandr}/bin/xrandr --output "$(${pkgs.xorg.xrandr}/bin/xrandr | ${pkgs.gawk}/bin/awk '/primary/ { print $1 }')" --brightness 1
+            '';
+          }
+          # Undim & lock after 10 more seconds
+          {
+            duration = 10;
+            command = ''
+              ${pkgs.xorg.xrandr}/bin/xrandr --output "$(${pkgs.xorg.xrandr}/bin/xrandr | ${pkgs.gawk}/bin/awk '/primary/ { print $1 }')" --brightness 1
+              ${pkgs.i3lock}/bin/i3lock
+            '';
+          }
+          # Finally, suspend an hour after it locks
+          {
+            duration = 3600;
+            command = ''
+              ${pkgs.systemd}/bin/systemctl suspend
+            '';
+          }
+        ];
+        # We use uniq here because conflicting definitions of the list of timers should
+        # be brought to the attention of the user.  Just appending one to the other
+        # usually would not make sense and definitely should not be done implicitly and
+        # silently.
+        type = types.uniq (types.listOf (types.submodule {
```

Comment:
In general, the list type should be avoided if possible. And in this case this can work well by just having to name the timers such that this can be `attrsOf submodule`. This then also won't clear timers specified with the compatibility options if users set them.

Reply:
It doesn't make sense to me to ever merge timer lists because the durations are relative -- inserting a timer entry delays all the timers after it.  For example, if you wish screens to lock after 10 minutes of inactivity and then wish to insert something before this entry that happens after 3 minutes of inactivity, you must *modify* the screen-lock timer's duration from 10 minutes to 7 minutes, because it is specified relative to the previous timer.

Contemplating a hypothetical for just a moment: We could make the Nix configuration use absolute timer durations and convert these to relative durations to pass them to xidlehook.  But this diverges from how xidlehook is normally used & would confuse users trying to move existing xidlehook configuration to NixOS.  Also, it opens up the question of how to merge cancellers.  If the Nth timer's action both undoes the previous timer's action and performs its own (as in upstream's xrandr-dimmer example) what happens when another timer is inserted between them?  One could imagine providing a merge-friendly interface that does stack unwinding, automatically adding each timer's canceller onto the next timer's action such that timers could be inserted at any point and still have everything chained correctly.  This is now mergable, but it is very different than the normal semantics of xidlehook.

Instead, I think it makes sense to treat one timer-list as one opaque value.  If two timer-lists are specified, this should raise an error that can be resolved by choosing one of them with mkForce.